sbi/inference/snpe/snpe_c.py

```diff
@@ -231,5 +370,354 @@ def _log_prob_proposal_posterior(
+
+        means_pp = self._means_proposal_posterior(
+            covariances_pp, means_p, precisions_p, means_d, precisions_d,
+        )
+
+        logits_pp = self._logits_proposal_posterior(
+            means_pp,
+            precisions_pp,
+            covariances_pp,
+            logits_p,
+            means_p,
+            precisions_p,
+            logits_d,
+            means_d,
+            precisions_d,
+        )
+
+        return logits_pp, means_pp, precisions_pp, covariances_pp
+
+    def _precisions_proposal_posterior(
+        self, precisions_p: Tensor, precisions_d: Tensor,
+    ):
+        """
+        Return the precisions and covariances of the proposal posterior.
+
+        Args:
+            precisions_p: Precision matrices of the proposal distribution.
+            precisions_d: Precision matrices of the density estimator.
+
+        Returns: (Precisions, Covariances) of the proposal posterior. L*K terms.
+        """
+
+        num_comps_p = precisions_p.shape[1]
+        num_comps_d = precisions_d.shape[1]
+
+        precisions_p_rep = precisions_p.repeat_interleave(num_comps_d, dim=1)
+        precisions_d_rep = precisions_d.repeat(1, num_comps_p, 1, 1)
+
+        precisions_pp = precisions_p_rep + precisions_d_rep
+        if isinstance(self._maybe_z_scored_prior, mvn.MultivariateNormal):
+            precisions_pp -= self._maybe_z_scored_prior.precision_matrix
+
+        covariances_pp = torch.inverse(precisions_pp)
+
+        return precisions_pp, covariances_pp
+
+    def _means_proposal_posterior(
+        self,
+        covariances_pp: Tensor,
+        means_p: Tensor,
+        precisions_p: Tensor,
+        means_d: Tensor,
+        precisions_d: Tensor,
+    ):
+        """
+        Return the means of the proposal posterior.
+
+        means_pp = C_ix * (P_i * m_i + P_x * m_x - P_o * m_o).
+
+        Args:
+            covariances_pp: Covariance matrices of the proposal posterior.
+            means_p: Means of the proposal distribution.
+            precisions_p: Precision matrices of the proposal distribution.
+            means_d: Means of the density estimator.
+            precisions_d: Precision matrices of the density estimator.
+
+        Returns: Means of the proposal posterior. L*K terms.
+        """
+
+        num_comps_p = precisions_p.shape[1]
+        num_comps_d = precisions_d.shape[1]
+
+        # First, compute the product P_i * m_i and P_j * m_j
+        cov_m_prod_p = _batched_mv(precisions_p, means_p)
+        cov_m_prod_d = _batched_mv(precisions_d, means_d)
+
+        # Repeat them to allow for matrix operations: same trick as for the precisions.
+        cov_m_prod_p_rep = cov_m_prod_p.repeat_interleave(num_comps_d, dim=1)
+        cov_m_prod_d_rep = cov_m_prod_d.repeat(1, num_comps_p, 1)
+
+        # Means = C_ij * (P_i * m_i + P_j * m_j).
+        summed_cov_m_prod_rep = cov_m_prod_p_rep + cov_m_prod_d_rep
+        if isinstance(self._maybe_z_scored_prior, mvn.MultivariateNormal):
```

Comment:
can we guarantee that this prior is exactly one of MVN or Uniform?

Reply:
If it is not, we will not use non-atomic loss. Instead just fall back on atomic loss. (`self.use_non_atomic_loss` will be set to `False` at the `__call__()` in `snpe_c.py`)